tensorflow_probability/examples/keras_bayesian_neural_network.py

```diff
@@ -0,0 +1,360 @@
+    train_seq = MNISTSequence(batch_size=FLAGS.batch_size,
+                              size=NUM_TRAIN_EXAMPLES)
+    heldout_seq = MNISTSequence(batch_size=FLAGS.batch_size,
+                                size=NUM_HELDOUT_EXAMPLES)
+  else:
+    train_set, heldout_set = tf.keras.datasets.mnist.load_data()
+    train_seq = MNISTSequence(data=train_set, batch_size=FLAGS.batch_size)
+    heldout_seq = MNISTSequence(data=heldout_set, batch_size=FLAGS.batch_size)
+
+  model = create_model()
+
+  print(" ... Training convolutional neural network")
+  for epoch in range(FLAGS.num_epochs):
+    epoch_accuracy = []
+    for step, (batch_x, batch_y) in enumerate(train_seq):
+      # Eager mode returns a Tensor objest and not a scalar
+      # value for the loss, therefore only the mean accuracy
+      # is displayed.
+      batch_accuracy = model.train_on_batch(
+          batch_x, batch_y)[1]
+      epoch_accuracy.append(batch_accuracy)
+
+      if step % 100 == 0:
+        print('Epoch: {}, Batch index: {}, Accuracy: {:.3f}'.format(
+            epoch, step, np.mean(epoch_accuracy)))
+
+      if (step+1) % FLAGS.viz_steps == 0:
+        # Compute log prob of heldout set by averaging draws from the model:
+        # p(heldout | train) = int_model p(heldout|model) p(model|train)
+        #                   ~= 1/n * sum_{i=1}^n p(heldout | model_i)
+        # where model_i is a draw from the posterior p(model|train).
+        print(" ... Running monte carlo inference")
+        probs = np.asarray([model.predict_generator(heldout_seq, verbose=1)
+                            for _ in range(FLAGS.num_monte_carlo)])
+        mean_probs = np.mean(probs, axis=0)
+        heldout_log_prob = np.mean(np.log(mean_probs))
+        print(" ... Held-out nats: {:.3f}".format(heldout_log_prob))
+
+        if HAS_SEABORN:
+          names = [layer.name for layer in model.layers
+                   if 'flipout' in layer.name]
+          qm_vals = [layer.kernel_posterior.mean()
+                     for layer in model.layers
+                     if 'flipout' in layer.name]
+          qs_vals = [layer.kernel_posterior.stddev()
+                     for layer in model.layers
+                     if 'flipout' in layer.name]
+          plot_weight_posteriors(names, qm_vals, qs_vals,
+                                 fname=os.path.join(
+                                     FLAGS.model_dir,
+                                     "epoch{}_step{:05d}_weights.png".format(
+                                         epoch, step)))
+          plot_heldout_prediction(heldout_seq.images, probs,
+                                  fname=os.path.join(
+                                      FLAGS.model_dir,
+                                      "epoch{}_step{}_pred.png".format(
+                                          epoch, step)),
+                                  title="mean heldout logprob {:.2f}"
+                                  .format(heldout_log_prob))
```

Comment:
Why don't use Keras' `fit` method? The code still looks unnecessarily verbose. I had actually updated this example to TF2, but I hadn't opened a pull request.

Reply:
The main reason for using the `train_on_batch` instead of the `fit` was because I am trying to preserve the operations of the original example, as I think these add some valuable insight on how bayesian networks are trained and their behaviour regarding uncertainty. I thought that this was the easier way to implement those instead of using some custom Keras callbacks, which I think would be more difficult to follow. When I first submitted the PR this example was working fine, still works for version 2.0. But since version 2.1 I am also getting the problem you described in issue #620. 

@davmre Is there any way you might be able to point me to a starting point for dealing with issue #620. I am making the naive assumption that the fix should not be too complicated, since as @nbro says the problem appears to be caused by the Convolution2DFlipout layers (specifically the kernel divergence) and not the DenseFlipout layers.

Reply:
I don't agree with you. I think that, as a rule of thumb, people should use `fit` and callbacks are perfectly fine and intuitive. If you decide to use `fit`, you can solve that issue by following the instructions here: https://github.com/tensorflow/tensorflow/issues/33729#issuecomment-562656807.

Reply:
@nbro I am not particularly fond of this solution because I can similarly solve the issue by disabling eager execution. In both cases the underlying cause is still not treated. However, I will make the changes you suggest (both using the `fit` function and callbacks) if @davmre also agrees with that implementation.

Reply:
"In both cases the underlying cause is still not treated.", because this is a bug in TensorFlow! Hopefully, someone is already taking care of this issue. `experimental_run_tf_function`, in fact, exists because there's an experimental feature related to distributed training.